pkg/tarbackfill/tarbackfill.go

```diff
@@ -0,0 +1,159 @@
+package tarbackfill
+
+import (
+	"archive/tar"
+	"io"
+	"path"
+	"strings"
+)
+
+// Reader wraps a tar.Reader so that if an item which would be read from it is
+// in a directory which is not included in the archive, a specified Backfiller
+// interface's Backfill() method will be called to supply a tar.Header which
+// will be inserted into the stream just ahead of that item.
+type Reader struct {
+	*tar.Reader
+	backfiller      Backfiller
+	seen            map[string]struct{}
```

Comment:
(Some documentation, or maybe naming this `seenDirs`, would be helpful for readers.)